gradle-jdks/src/main/java/com/palantir/gradle/jdks/GradleWrapperPatcher.java

```diff
@@ -0,0 +1,182 @@
+/*
+ * (c) Copyright 2024 Palantir Technologies Inc. All rights reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.palantir.gradle.jdks;
+
+import com.palantir.gradle.autoparallelizable.AutoParallelizable;
+import java.io.File;
+import java.io.FileInputStream;
+import java.io.FileNotFoundException;
+import java.io.IOException;
+import java.io.InputStream;
+import java.nio.charset.StandardCharsets;
+import java.nio.file.Files;
+import java.nio.file.Path;
+import java.util.ArrayList;
+import java.util.List;
+import java.util.Optional;
+import java.util.OptionalInt;
+import java.util.Properties;
+import java.util.stream.Collectors;
+import java.util.stream.IntStream;
+import javax.inject.Inject;
+import org.apache.commons.io.IOUtils;
+import org.gradle.api.file.FileCollection;
+import org.gradle.api.file.ProjectLayout;
+import org.gradle.api.logging.Logger;
+import org.gradle.api.logging.Logging;
+
+@AutoParallelizable
+public abstract class GradleWrapperPatcher {
+
+    private static final Logger log = Logging.getLogger(GradleWrapperPatcher.class);
+    private static final String ENABLE_GRADLE_JDK_SETUP = "gradle.jdk.setup.enabled";
+    private static final String GRADLEW_PATCH = "gradlew-patch";
+    private static final String GRADLEW_UNIX_SCRIPT = "gradlew";
+    private static final String COMMENT_BLOCK = "###";
+    private static final String SHEBANG = "#!";
+
+    // DO NOT CHANGE the header and the footer, they are used to identify the patch block
+    private static final String GRADLEW_PATCH_HEADER = "# >>> Gradle JDK setup >>>";
+    private static final String GRADLEW_PATCH_FOOTER = "# <<< Gradle JDK setup <<<";
+
+    interface Params {
+        @Inject
+        ProjectLayout getProjectLayout();
+    }
+
+    public abstract static class GradleWrapperPatcherTask extends GradleWrapperPatcherTaskImpl {}
+
+    static void action(Params params) {
+        if (!isGradleJdkSetupEnabled(params.getProjectLayout())) {
+            return;
+        }
+        log.lifecycle("Gradle JDK setup is enabled, patching the gradle wrapper files");
+        Path gradlewFile = params.getProjectLayout()
+                .files(GRADLEW_UNIX_SCRIPT)
```

Comment:
We should pass this in as a `@OutputFile RegularFileProperty getGradlewScript();`. I think it's not possible to make something both an input and output (worth trying though), so you may need to set `.upToDateWhen { false }` in the `GradleWrapperPatcherTask` constructor.